bot/cogs/moderation/scheduler.py

```diff
@@ -415,4 +415,6 @@ def mod_log(self) -> ModLog:
         expiry = dateutil.parser.isoparse(infraction["expires_at"]).replace(tzinfo=None)
         await time.wait_until(expiry)
 
-        await self.deactivate_infraction(infraction)
+        # Because deactivate_infraction() explicitly cancels this scheduled task, it runs in
+        # a separate task to avoid prematurely cancelling itself.
+        self.bot.loop.create_task(self.deactivate_infraction(infraction))
```

Comment:
I think this is decent, but is there a better workaround? My other idea was to take out the `cancel_task()` from `deactivate_infraction()` and have the other callers of `deactivate_infraction()` be responsible for cancelling the task. I think that's a worse solution than what's above.

Reply:
Maybe I just don't follow the flow of execution, but if you're afraid that a scheduled task is being cancelled because of its parent being cancelled, you could also use [`asyncio.shield`](https://docs.python.org/3/library/asyncio-task.html#asyncio.shield) to prevent it from being cancelled in such a situation.

Reply:
The scheduler base class holds onto a `Task` for `_scheduled_task()`. Inside `_scheduled_task()`, the coroutine `deactivate_infraction()` is awaited. Inside this coroutine, the aforementioned task is cancelled. 

https://github.com/python-discord/bot/blob/f09d6b0646edb62806b60b145986b7cc680fd77c/bot/cogs/moderation/scheduler.py#L381

So a child is effectively cancelling its parent before the child itself can finish. Hope that makes sense. Does shield work to prevent that situation? I'm not familiar with it.

Reply:
I am realising there is a side issue here when an infraction gets edited and re-scheduled in the middle of that same infraction being deactivated. Still, it's good that a deactivation doesn't get aborted half-way (could leave it in a poor state e.g. expired in the DB but the user still has the mute role).

Reply:
Was just being consistent with the rest of the code but you're right, we should start taking opportunities to phase out `loop.whatever` where applicable.

Anyway, this has been replaced with a shield as you suggest so it's no longer relevant.